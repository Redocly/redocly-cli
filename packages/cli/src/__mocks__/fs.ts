export const existsSync = jest.fn();
export const readFileSync = jest.fn(() => '');
export const statSync = jest.fn(() => ({ size: 0 }));
export const createReadStream = jest.fn();
export const writeFileSync = jest.fn();
export const mkdirSync = jest.fn();
