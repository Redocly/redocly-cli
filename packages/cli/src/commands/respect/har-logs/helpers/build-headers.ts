/**
 * Support the three possible header formats we'd get from a request or
 * response:
 *
 * - A flat array with both names and values: [name, value, name, value, ...]
 * - An object with array values: { name: [value, value] }
 * - An object with string values: { name: value }
 */
export function buildHeaders(headers: any): any[] {
  const list = [];
  if (Array.isArray(headers)) {
    for (let i = 0; i < headers.length; i += 2) {
      list.push({
        name: headers[i],
        value: headers[i + 1],
      });
    }
  } else if (headers instanceof Map || headers.entries) {
    // Handle both Map and Headers objects (which have entries())
    for (const [name, value] of headers.entries()) {
      list.push({ name, value });
    }
  } else if (typeof headers === 'object') {
    for (const [name, value] of Object.entries(headers)) {
      list.push({ name, value: value as string });
    }
  }
  return list;
}
