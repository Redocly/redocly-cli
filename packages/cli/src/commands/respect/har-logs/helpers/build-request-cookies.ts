import * as cookie from 'cookie';

export function buildRequestCookies(headers: any): any[] {
  const cookies: any[] = [];
  for (const [header, headerValue] of Object.entries(headers)) {
    if (header.toLowerCase() === 'cookie') {
      // Handle both array and single string cases
      const cookieValues = Array.isArray(headerValue) ? headerValue : [headerValue];
      for (const value of cookieValues) {
        const parsed = cookie.parse(value);
        for (const [name, value] of Object.entries(parsed)) {
          cookies.push({ name, value });
        }
      }
    }
  }
  return cookies;
}
