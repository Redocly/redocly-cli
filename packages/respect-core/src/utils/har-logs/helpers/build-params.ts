import * as querystring from 'querystring';

export function buildParams(paramString: string): any[] {
  const params = [];
  const parsed = querystring.parse(paramString);
  for (const name in parsed) {
    const value = parsed[name];
    if (Array.isArray(value)) {
      value.forEach((item) => {
        params.push({ name, value: item });
      });
    } else {
      params.push({ name, value });
    }
  }
  return params;
}
