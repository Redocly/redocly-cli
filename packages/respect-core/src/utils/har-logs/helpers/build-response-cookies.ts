import * as setCookie from 'set-cookie-parser';

export function buildResponseCookies(headers: any): any[] {
  const cookies: any[] = [];
  // Handle both plain objects and Headers instance
  const setCookies =
    headers instanceof Headers
      ? headers.getSetCookie() // For Headers object
      : headers['set-cookie']; // For plain objects

  if (setCookies) {
    for (const headerValue of setCookies) {
      let parsed;
      try {
        parsed = setCookie.parse(headerValue);
      } catch {
        continue;
      }
      for (const cookie of parsed) {
        const { name, value, path, domain, expires, httpOnly, secure } = cookie;
        const harCookie: any = {
          name,
          value,
          httpOnly: httpOnly || false,
          secure: secure || false,
        };
        if (path) {
          harCookie.path = path as string | undefined;
        }
        if (domain) {
          harCookie.domain = domain as string | undefined;
        }
        if (expires) {
          harCookie.expires = expires.toISOString();
        }
        cookies.push(harCookie);
      }
    }
  }
  return cookies;
}
