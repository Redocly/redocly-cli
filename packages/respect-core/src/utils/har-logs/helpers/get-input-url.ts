export function getInputUrl(input: any): URL {
  const url = typeof input === 'string' ? input : input.url;
  return new URL(url);
}
