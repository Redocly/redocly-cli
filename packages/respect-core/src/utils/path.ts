import { isBrowser } from '@redocly/openapi-core';

let pathLib: any;

if (isBrowser) {
  // Browser environment
  pathLib = await import('path-browserify');
} else {
  // Node.js environment
  pathLib = await import('node:path');
}

export default pathLib;
