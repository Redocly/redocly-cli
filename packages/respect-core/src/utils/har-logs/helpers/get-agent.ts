import * as http from 'http';
import * as https from 'https';
import { Dispatcher } from 'undici';
import { getInputUrl } from './get-input-url';
import { handleRequest } from './handle-request';

const HarHttpAgent = createAgentClass(http.Agent);
const HarHttpsAgent = createAgentClass(https.Agent);

let globalHttpAgent: any;
let globalHttpsAgent: any;

// Add new Undici dispatcher
class HarDispatcher extends Dispatcher {
  constructor(opts?: any) {
    super(opts);
  }

  dispatch(options: Dispatcher.DispatchOptions, handler: Dispatcher.DispatchHandlers): boolean {
    // Handle HAR logging here similar to handleRequest
    handleRequest({ input: options, handler, harLog: new Map(), isUndici: true });
    return super.dispatch(options, handler);
  }
}

export function getAgent(input: any, options: any): any {
  // Add Undici dispatcher support
  if (options.dispatcher) {
    if (options.dispatcher instanceof Dispatcher) {
      return new HarDispatcher(options.dispatcher);
    }
    return options.dispatcher;
  }

  if (options.agent) {
    if (typeof options.agent === 'function') {
      return function (...args: any[]) {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        const agent = options.agent.call(this, ...args);
        if (agent) {
          instrumentAgentInstance(agent);
          return agent;
        }
        return getGlobalAgent(input);
      };
    }
    instrumentAgentInstance(options.agent);
    return options.agent;
  }
  return getGlobalAgent(input);
}

function getGlobalAgent(input: any): any {
  const url = getInputUrl(input);
  if (url.protocol === 'http:') {
    if (!globalHttpAgent) {
      globalHttpAgent = new HarHttpAgent();
    }
    return globalHttpAgent;
  }
  if (!globalHttpsAgent) {
    globalHttpsAgent = new HarHttpsAgent();
  }
  return globalHttpsAgent;
}

/**
 * Instrument an existing Agent instance. This overrides the instance's
 * `addRequest` method. It should be fine to continue using for requests made
 * without `withHar` - if the request doesn't have our `x-har-request-id`
 * header, it won't do anything extra.
 */
function instrumentAgentInstance(agent: any): void {
  const { addRequest: originalAddRequest } = agent;
  if (!originalAddRequest.isHarEnabled) {
    agent.addRequest = function addRequest(request: any, ...args: []) {
      handleRequest(request, ...args);
      return originalAddRequest.call(this, request, ...args);
    };
    agent.addRequest.isHarEnabled = true;
  }
}

function createAgentClass(BaseAgent: any): any {
  class HarAgent extends BaseAgent {
    constructor(...args: any[]) {
      super(...args);
      (this.addRequest as any).isHarEnabled = true;
    }

    addRequest(request: any, ...args: []): void {
      handleRequest(request, ...args);
      return super.addRequest(request, ...args);
    }
  }

  return HarAgent;
}
