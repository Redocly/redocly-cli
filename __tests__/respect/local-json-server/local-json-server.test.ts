import { spawn } from 'child_process';
import { join } from 'node:path';
import * as fs from 'node:fs';
import { getCommandOutput, getParams } from '../../helpers';

const dbPath = join(__dirname, 'fake-db.json');

describe('local-json-server', () => {
  let serverProcess: any;
  let originalData: string | undefined;

  beforeAll(async () => {
    // Start json-server
    serverProcess = spawn('npm', ['run', 'json-server'], { detached: true });

    // Store original state of fake-bd.json
    originalData = fs.readFileSync(dbPath, 'utf8');
  });

  afterAll(() => {
    // Kill the process group to ensure child processes are cleaned up
    process.kill(-serverProcess.pid);

    // Restore original state
    if (originalData) {
      fs.writeFileSync(dbPath, originalData);
    }
  });

  test('local-json-server test case', () => {
    const indexEntryPoint = join(process.cwd(), 'packages/cli/lib/index.js');
    const fixturesPath = join(__dirname, 'local-json-server.arazzo.yaml');
    const args = getParams(indexEntryPoint, ['respect', fixturesPath]);

    const result = getCommandOutput(args);
    expect(result).toMatchSnapshot();
  });
});
