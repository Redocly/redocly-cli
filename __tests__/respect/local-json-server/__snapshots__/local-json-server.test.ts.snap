// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`local-json-server local-json-server test case 1`] = `
"────────────────────────────────────────────────────────────────────────────────

  Running workflow local-json-server.arazzo.yaml / before

  ✓ GET /items - step get-all-items-before
    ✓ status code check - $statusCode in [200]
    ✓ content-type check
    ✓ schema check

────────────────────────────────────────────────────────────────────────────────

  Running workflow local-json-server.arazzo.yaml / crud-items

  ✓ POST /items - step add-item
    ✓ success criteria check - $statusCode == 201
    ✓ status code check - $statusCode in [201]
    ✓ content-type check
    ✓ schema check

  ✗ GET /items - step get-all-items-again
    ✗ success criteria check - $response.body#/0/value == 100
    ✓ status code check - $statusCode in [200]
    ✓ content-type check
    ✓ schema check

────────────────────────────────────────────────────────────────────────────────

  Running workflow local-json-server.arazzo.yaml / auto-inherit

  ✓ POST /items - step add-item-auto-inherit
    ✓ status code check - $statusCode in [201]
    ✓ content-type check
    ✓ schema check

  ✓ GET /items - step get-all-items
    ✓ status code check - $statusCode in [200]
    ✓ content-type check
    ✓ schema check

  ✓ GET /items/{id} - step get-item-by-id-auto-inherit
    ✓ status code check - $statusCode in [200, 404]
    ✓ content-type check

  ✓ DELETE /items/{id} - step drop-item-auto-inherit
    ✓ success criteria check - $statusCode == 200
    ✓ status code check - $statusCode in [200]


  Failed tests info:

  Workflow name: crud-items

    stepId - get-all-items-again
    ✗ success criteria check
      Checking simple criteria: {"condition":"$response.body#/0/value == 100"}
      
  Summary for local-json-server.arazzo.yaml
  
  Workflows: 2 passed, 1 failed, 3 total
  Steps: 6 passed, 1 failed, 7 total
  Checks: 20 passed, 1 failed, 21 total
  Time: <test>ms


┌───────────────────────────────────────────────────────────────────────┬────────────┬─────────┬─────────┬──────────┐
│ Filename                                                              │ Workflows  │ Passed  │ Failed  │ Warnings │
├───────────────────────────────────────────────────────────────────────┼────────────┼─────────┼─────────┼──────────┤
│ x local-json-server.arazzo.yaml                                       │ 3          │ 2       │ 1       │ -        │
└───────────────────────────────────────────────────────────────────────┴────────────┴─────────┴─────────┴──────────┘



 Tests exited with error 
"
`;
